posts/2024-06-05-mixing-python-envs.md

```diff
@@ -0,0 +1,181 @@
+
+## Virtual environments
+Instead, we recommend creating a [virtual environment](https://docs.python.org/3/library/venv.html). A virtual environment is a self-contained Python
+environment built on top of an existing Python installation. The self-contained
+nature of a virtual environment means that any issues due to conflicting 
+dependencies can be avoided entirely. Being able to base the environment on top of
+an existing `conda/analysis3` environment means that the installation can be kept
+small, as most of the dependencies will have already been satisfied. 
+
+There are only two additional commands required to install `xWMB` into a virtual environment, as shown below:
+<terminal-window typingDelay=30 lineDelay=200>
+    <terminal-line data="input">module use /g/data/hh5/public/modules</terminal-line>
+    <terminal-line data="input">module load conda/analysis3</terminal-line>
+    <terminal-line data="input">python3 -m venv xwmb_venv -\\\-system-site-packages</terminal-line>
+    <terminal-line data="input">source xwmb_venv/bin/activate</terminal-line>
+    <terminal-line data="input" inputChar="(xwmb_venv) $">pip install git+https://github.com/hdrake/xwmb.git@main</terminal-line>
+    <terminal-line>Collecting git+https://github.com/hdrake/xwmb.git@main</terminal-line>
+    <terminal-line>  Cloning https://github.com/hdrake/xwmb.git (to revision main) to /scratch/v45/dr4292/tmp/pip-req-build-qox565pe</terminal-line>
+    <terminal-line>  Running command git clone -\\\-filter=blob:none -\\\-quiet https://github.com/hdrake/xwmb.git /scratch/v45/dr4292/tmp/pip-req-build-qox565pe</terminal-line>
+    <terminal-line>  ...</terminal-line>
+    <terminal-line>Successfully installed regionate-0.0.1 sectionate-0.2.1 xbudget-0.1.0 xgcm-0.8.2.dev15+g7492277 xwmb-0.1.1 xwmt-0.1.1</terminal-line>
+</terminal-window>
+```{warning}
+Do not use `pip install --user` here, as this will result in the packages being installed in your `~/.local` directory, leaving the virtual environment empty
+```
+
+And that's it. We now have a fully self-contained environment with a custom 
+package built on top of the `conda/analysis3` environment. It is important to note
+that this virtual environment is tied to the version of `conda/analysis3` that was loaded
+at the time of its creation, so it avoids the out-of-date dependencies issue mentioned earlier.
+<terminal-window typingDelay=30 lineDelay=200>
+    <terminal-line data="input" inputChar="(xwmb_venv) $">which python3</terminal-line>
+    <terminal-line>~/xwmb_venv/bin/python3</terminal-line>
+    <terminal-line data="input" inputChar="(xwmb_venv) $">ls -l ~/xwmb_venv/bin/python3</terminal-line>
+    <terminal-line>lrwxrwxrwx 1 dr4292 v45 67 Jun  5 14:43 /home/563/dr4292/xwmb_venv/bin/python3 -> /g/data/hh5/public/apps/miniconda3/envs/analysis3-24.01/bin/python3</terminal-line>
+</terminal-window>
+
+There are a few different ways that this virtual environment can be used in scripts.
```

Comment:
it's not clear which are the different ways , I would remove the reference to the shebang line in the first sentence as it can confuse inexperienced people. Something like:
Immediately after loading the `conda/analysis3` module, run 
`source xwmb_venv/bin/activate` and any `python` command or 
script will now run under the new virtual environment.
If you're running a script directly, you can
change the path in the shebang line to the path to the `python` symlink in the virtual environment:...

You can then add the shebang link in the second sentence.

Reply:
If you want to be more specific and show a script example just do it as you're doing for the shebang line.